File: src/cli/commands/report.ts
import { existsSync, readFileSync } from 'node:fs';
import { join } from 'node:path';
import type { EngineResult } from '../../core/engine.js';
import { fail } from '../format.js';
import type { CommandResult } from '../types.js';
import { STATE_DIR } from './record.js';
import { SCORES_STATE_FILENAME } from './score.js';

export type ReportFormat = 'table' | 'json' | 'markdown';

export interface ReportOptions {
  json: boolean;
  format: ReportFormat;
}

interface ReportRow {
  learnerId: string;
  skillId: string;
  model: string;
  metric: string;
  value: number;
  responseCount: number;
}

function buildRows(result: EngineResult): ReportRow[] {
  const rows: ReportRow[] = [];
  for (const report of result.reports) {
    for (const learner of report.learners) {
      for (const skill of learner.skills) {
        rows.push({
          learnerId: learner.learnerId,
          skillId: skill.skillId,
          model: report.model,
          metric: skill.metric,
          value: skill.value,
          responseCount: skill.responseCount,
        });
      }
    }
  }
  rows.sort(
    (a, b) =>
      a.learnerId.localeCompare(b.learnerId) ||
      a.skillId.localeCompare(b.skillId) ||
      a.model.localeCompare(b.model),
  );
  return rows;
}

const TABLE_HEADER = ['learner', 'skill', 'model', 'metric', 'value', 'responses'];

function renderTable(rows: ReportRow[]): string {
  if (rows.length === 0) {
    return 'No scores found.\n';
  }
  const data = rows.map((r) => [
    r.learnerId,
    r.skillId,
    r.model,
    r.metric,
    r.value.toFixed(4),
    String(r.responseCount),
  ]);
  const widths = TABLE_HEADER.map((h, i) =>
    Math.max(h.length, ...data.map((row) => (row[i] ?? '').length)),
  );
  const renderRow = (cells: string[]): string =>
    cells.map((c, i) => c.padEnd(widths[i] ?? 0)).join('  ');
  return (
    [renderRow(TABLE_HEADER), renderRow(widths.map((w) => '-'.repeat(w))), ...data.map(renderRow)].join(
      '\n',
    ) + '\n'
  );
}

function renderMarkdown(rows: ReportRow[]): string {
  if (rows.length === 0) {
    return 'No scores found.\n';
  }
  const lines = [
    `| ${TABLE_HEADER.join(' | ')} |`,
    `| ${TABLE_HEADER.map(() => '---').join(' | ')} |`,
    ...rows.map(
      (r) =>
        `| ${r.learnerId} | ${r.skillId} | ${r.model} | ${r.metric} | ${r.value.toFixed(4)} | ${r.responseCount} |`,
    ),
  ];
  return lines.join('\n') + '\n';
}

/**
 * Reads `.masterytrace/scores.json` (written by `masterytrace score`) and
 * prints a per-learner, per-skill mastery table in the requested format.
 * The global `--json` flag always wins over `--format` and prints the raw
 * stored report as machine-readable JSON.
 */
export function runReport(cwd: string, options: ReportOptions): CommandResult {
  const scoresPath = join(cwd, STATE_DIR, SCORES_STATE_FILENAME);
  if (!existsSync(scoresPath)) {
    const message = `No scores found at ${scoresPath}. Run 'masterytrace score' first.`;
    return fail(1, options.json, { error: message }, `${message}\n`);
  }

  let result: EngineResult;
  try {
    result = JSON.parse(readFileSync(scoresPath, 'utf-8')) as EngineResult;
  } catch (error) {
    const message = error instanceof Error ? error.message : String(error);
    return fail(1, options.json, { error: message }, `Error reading scores: ${message}\n`);
  }

  if (options.json) {
    return { exitCode: 0, stdout: `${JSON.stringify(result)}\n` };
  }

  const rows = buildRows(result);
  let text: string;
  if (options.format === 'markdown') {
    text = renderMarkdown(rows);
  } else if (options.format === 'json') {
    text = `${JSON.stringify(result, null, 2)}\n`;
  } else {
    text = renderTable(rows);
  }

  return { exitCode: 0, stdout: text };
}

File: python/src/masterytrace/cli/commands/report.py
"""
Reads `.masterytrace/scores.json` (written by `masterytrace score`) and
prints a per-learner, per-skill mastery table in the requested format.
Ported from src/cli/commands/report.ts.
"""
from __future__ import annotations

import json
from dataclasses import dataclass
from pathlib import Path
from typing import Any, Dict, List

from ..format import fail
from ..types import CommandResult
from .record import STATE_DIR
from .score import SCORES_STATE_FILENAME

ReportFormat = str  # "table" | "json" | "markdown"


@dataclass
class ReportOptions:
    json: bool
    format: ReportFormat


def _build_rows(result: Dict[str, Any]) -> List[Dict[str, Any]]:
    rows: List[Dict[str, Any]] = []
    for report in result.get("reports", []):
        for learner in report.get("learners", []):
            for skill in learner.get("skills", []):
                rows.append(
                    {
                        "learner_id": learner["learner_id"],
                        "skill_id": skill["skill_id"],
                        "model": report["model"],
                        "metric": skill["metric"],
                        "value": skill["value"],
                        "response_count": skill["response_count"],
                    }
                )
    rows.sort(key=lambda r: (r["learner_id"], r["skill_id"], r["model"]))
    return rows


_TABLE_HEADER = ["learner", "skill", "model", "metric", "value", "responses"]


def _render_table(rows: List[Dict[str, Any]]) -> str:
    if not rows:
        return "No scores found.\n"
    data = [
        [r["learner_id"], r["skill_id"], r["model"], r["metric"], f"{r['value']:.4f}", str(r["response_count"])]
        for r in rows
    ]
    widths = [max(len(_TABLE_HEADER[i]), *(len(row[i]) for row in data)) for i in range(len(_TABLE_HEADER))]

    def render_row(cells: List[str]) -> str:
        return "  ".join(cell.ljust(widths[i]) for i, cell in enumerate(cells))

    lines = [render_row(_TABLE_HEADER), render_row(["-" * w for w in widths])]
    lines.extend(render_row(row) for row in data)
    return "\n".join(lines) + "\n"


def _render_markdown(rows: List[Dict[str, Any]]) -> str:
    if not rows:
        return "No scores found.\n"
    lines = [
        f"| {' | '.join(_TABLE_HEADER)} |",
        f"| {' | '.join('---' for _ in _TABLE_HEADER)} |",
    ]
    for r in rows:
        lines.append(
            f"| {r['learner_id']} | {r['skill_id']} | {r['model']} | {r['metric']} | {r['value']:.4f} | {r['response_count']} |"
        )
    return "\n".join(lines) + "\n"


def run_report(cwd: str, options: ReportOptions) -> CommandResult:
    scores_path = Path(cwd) / STATE_DIR / SCORES_STATE_FILENAME
    if not scores_path.exists():
        message = f"No scores found at {scores_path}. Run 'masterytrace score' first."
        return fail(1, options.json, {"error": message}, f"{message}\n")

    try:
        result = json.loads(scores_path.read_text(encoding="utf-8"))
    except Exception as error:  # noqa: BLE001 -- mirrors the TS catch-all for parse errors
        return fail(1, options.json, {"error": str(error)}, f"Error reading scores: {error}\n")

    if options.json:
        return CommandResult(exit_code=0, stdout=f"{json.dumps(result)}\n")

    rows = _build_rows(result)
    if options.format == "markdown":
        text = _render_markdown(rows)
    elif options.format == "json":
        text = json.dumps(result, indent=2) + "\n"
    else:
        text = _render_table(rows)

    return CommandResult(exit_code=0, stdout=text)
